enhancements/per-node-configuration.md

```diff
@@ -0,0 +1,484 @@
+
+### Alternative 1: Neighbor-Level Node Selectors
+
+Add node selectors directly to the `Neighbor` type within a single Underlay:
+
+```go
+type Neighbor struct {
+    // NodeSelector specifies which nodes should peer with this neighbor
+    // +optional
+    NodeSelector *metav1.LabelSelector `json:"nodeSelector,omitempty"`
+
+    ASN     uint32  `json:"asn,omitempty"`
+    Address string  `json:"address"`
+    // ... rest of fields
+}
+```
+
+**Example:**
+```yaml
+apiVersion: openpe.openperouter.github.io/v1alpha1
+kind: Underlay
+metadata:
+  name: underlay
+spec:
+  asn: 64512
+  nics:
+    - toswitch
+  neighbors:
+    - asn: 64500
+      address: 192.168.1.254
+      nodeSelector:
+        matchLabels:
+          topology.kubernetes.io/rack: rack-1
+    - asn: 64500
+      address: 192.168.2.254
+      nodeSelector:
+        matchLabels:
+          topology.kubernetes.io/rack: rack-2
+```
+
+**Rejected because**:
+- Cannot handle per-node NIC name differences
+- Cannot support per-node ASN requirements
+- Less flexible for complex topologies
+- More complex to manage when multiple parameters vary per-node
+- Doesn't align with Kubernetes patterns (e.g., DaemonSet uses pod-level selectors, not container-level)
+
+## Implementation Plan
+
+### Phase 1: API Changes and Validation
+
+**Deliverables:**
+- Add `NodeSelector` field to `UnderlaySpec` (api/v1alpha1/underlay_types.go:23)
+- Update CRD generation (`make manifests`)
+- Implement validation webhook for overlapping selector detection
+- Add `MatchedNodes` field to `UnderlayStatus`
+- Add unit tests for validation logic
+
+**Testing:**
+- Unit tests for selector validation
+- Webhook tests for conflict detection
+- API validation tests for backward compatibility
+
+### Phase 2: Controller Enhancement
+
+**Deliverables:**
+- Add Node watch to Underlay controller, caching only the node where the
+  daemonset pod is running.
+- Implement node matching logic using `metav1.LabelSelector`
+- Update FRR configuration generation to be node-aware
+- Implement status reporting for matched nodes
+- Handle dynamic node label changes
+
+**Testing:**
+- Controller unit tests for node matching
+- Integration tests for configuration application
+- Tests for node label change handling
+- Tests for node addition/removal
+
+### Phase 3: Status Reporting
+
+**Deliverables:**
+- Populate `MatchedNodes` in status
+- Add Kubernetes conditions (Ready, NodesMatched, PartiallyConfigured)
+- Update status when node set changes
+
+**Testing:**
+- Status reporting tests
+- Condition transition tests
+
+### Phase 4: Documentation and Examples
+
+**Deliverables:**
+- Update API documentation
+- Create example configurations for:
+  - Multi-rack deployments
+  - Multi-datacenter deployments
+  - Mixed hardware deployments
+- Migration guide from cluster-wide to per-node configuration
+- Troubleshooting guide for selector conflicts
+
+### Phase 5: E2E Testing
+
+**Deliverables:**
+- E2E test for multi-rack scenario
```

Comment:
what does this look like ? a topology diagram would be incredible, but I just need to understand what does this mean at this point.

Reply:
I suppose a pair of Underlays for different nodes with different neighbors ASNs,